src/main/java/uk/gov/hmcts/reform/finrem/caseorchestration/service/documentchecker/DuplicateFilenameDocumentChecker.java

```diff
@@ -0,0 +1,125 @@
+package uk.gov.hmcts.reform.finrem.caseorchestration.service.documentchecker;
+
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.reform.finrem.caseorchestration.model.ccd.CaseDocument;
+import uk.gov.hmcts.reform.finrem.caseorchestration.model.ccd.DocumentFileNameProvider;
+import uk.gov.hmcts.reform.finrem.caseorchestration.model.ccd.FinremCaseData;
+import uk.gov.hmcts.reform.finrem.caseorchestration.model.ccd.FinremCaseDetails;
+import uk.gov.hmcts.reform.finrem.caseorchestration.model.ccd.HasCaseDocument;
+import uk.gov.hmcts.reform.finrem.caseorchestration.model.ccd.UploadDocument;
+import uk.gov.hmcts.reform.finrem.caseorchestration.model.ccd.UploadGeneralDocument;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+import static java.util.Optional.ofNullable;
+
+@Component
+@Slf4j
+public class DuplicateFilenameDocumentChecker implements DocumentChecker {
+
+    private static final String WARNING = "A document with this filename already exists on the case";
+
+    @Override
+    public boolean canCheck(CaseDocument caseDocument) {
+        return true;
+    }
+
+    private boolean isDuplicateFilename(CaseDocument caseDocument, Supplier<List<DocumentFileNameProvider>> caseDocumentSupplier) {
+        return ofNullable(caseDocumentSupplier.get()).orElse(List.of(CaseDocument.builder().documentFilename("").build()))
+            .stream().anyMatch(d -> d.getDocumentFilename().equals(caseDocument.getDocumentFilename()));
+    }
+
+    private static void processList(List<?> list, List<DocumentFileNameProvider> allDocuments) {
+        if (list != null) {
+            for (Object item : list) {
+                if (item instanceof HasCaseDocument hasCaseDocument) {
+                    processHasCaseDocument(hasCaseDocument, allDocuments);
+                } else {
+                    log.warn("Ignored " + item.getClass().getName());
+                }
+            }
+        }
+    }
+
```

Comment:
This is complex.  The method name is fine, but some JavaDoc for the method would be great.  Something that explains what the method does, to save the next dev from having to work it out :-D

Reply:
Java doc added.